File: Java/TenKindsOfPeople.java
import java.util.*;
import java.io.BufferedReader; 
import java.io.IOException; 
import java.io.InputStreamReader; 
import java.util.Scanner; 
import java.util.StringTokenizer;public class TenKindsOfPeople {
	public static void main(String[] args) throws IOException {
		  BufferedReader scan = new BufferedReader( 
                              new InputStreamReader(System.in));
		 StringBuilder output = new StringBuilder();
		String[] line = scan.readLine().split(" ");
		int r = Integer.parseInt(line[0]);
		int c = Integer.parseInt(line[1]);
		char[][] grid = new char[r][c];
		for (int i = 0; i < r; i++) {
			String l = scan.readLine();
			for (int j = 0; j < c; j++) {
				grid[i][j] = l.charAt(j);
			}
		}
		int n = Integer.parseInt(scan.readLine());
		for (int i = 0; i < n; i++) {
			String[] points = scan.readLine().split(" ");
			Point begin = new Point(Integer.parseInt(points[0]) - 1, Integer.parseInt(points[1]) - 1);
			Point end = new Point(Integer.parseInt(points[2]) - 1, Integer.parseInt(points[3]) - 1);
			char type = grid[begin.r][begin.c];
			if(DFS(grid, begin, r, c , end)) {
				if (type == '1'){
					output.append("decimal" + "\n");
				}
				else {
					output.append("binary" + "\n");
				}

				}
				else {
					output.append("neither" + "\n");
				}
			}
			System.out.println(output);

		}

	
	public static  boolean DFS(char[][] grid, Point start, int r, int c, Point finish) {
		char type = grid[start.r][start.c];
		Deque<Point> stack = new ArrayDeque<Point>();
		boolean[][] visited = new boolean[r][c];
		stack.addFirst(start);
		while (!stack.isEmpty()) {
			Point current = stack.removeFirst();
			if (current.r == finish.r && current.c == finish.c) {
				return true;
			}
			visited[current.r][current.c] = true;
			for (Point neighbor : current.getNeighors()) {
				if (isInBounds(neighbor.r, neighbor.c, r, c) && !visited[neighbor.r][neighbor.c] &&  grid[neighbor.r][neighbor.c] == (type)) {
					stack.addFirst(neighbor);
				}
			}
		}
		return false;
	}
	static boolean isInBounds(int r, int c, int rMax, int cMax) {
		return (r >= 0 &&
		c >= 0
		&& rMax > r
		&& cMax > c);

	}

}
class Point {
	int r;
	int c;
	String type;
	public Point(int i, int j) {
		r = i;
		c = j;
	}
	public ArrayList<Point> getNeighors() {
		ArrayList<Point> output = new ArrayList<Point>();
		output.add(new Point(r + 1, c));
		output.add(new Point(r - 1, c));
		output.add(new Point(r, c + 1));
		output.add(new Point(r, c - 1));
		return output;
	}
	
}

File: Python/tenkindsofpeople.py
def find(parents, child):
	if parents[child] == child:
		return child
	parents[child] = find(parents, parents[child])
	return parents[child]
def union(parents, childX, childY):
	parents[find(parents,childY)] = find(parents,childX)

def main():
	r, c = map(int, input().split())
	grid = list()
	connections = list(range(c * r))
	for i in range(r):
		grid.append(input())
	for i in range(r):
		for j in range(c):
			if i < r - 1 and grid[i][j] == grid[i + 1][j]:
				union(connections, i * c + j, (i + 1) * c + j)
			if j < c - 1 and grid[i][j] == grid[i][j + 1]:
				union(connections, i * c + j, i * c + j + 1)
	check = int(input())
	for i in range(check):
		r1, c1, r2, c2 = map(int, input().split())
		r1 -= 1
		c1 -= 1
		r2 -= 1
		c2 -= 1
		if  find(connections,r1 * c + c1) == find(connections, r2 * c + c2):
			if grid[r1][c1] == "1":
				print("decimal")
			else:
				print("binary")

		else:
			print("neither")

if __name__ == "__main__":
	main()

